desktop/renderer-app/src/pages/DeviceCheckPages/SpeakerCheckPage/index.tsx

```diff
@@ -4,71 +4,70 @@ import muteSVG from "../../../assets/image/mute.svg";
 import { DeviceSelect } from "../../../components/DeviceSelect";
 import { Device } from "../../../types/device";
-import { useRTCEngine } from "../../../utils/hooks/use-rtc-engine";
 import { DeviceCheckLayoutContainer } from "../DeviceCheckLayoutContainer";
 import { runtime } from "../../../utils/runtime";
 import { routeConfig } from "../../../route-config";
 import { DeviceCheckResults } from "../utils";
 import { useTranslation } from "react-i18next";
 import path from "path";
+import { FlatRTCContext } from "../../../components/FlatRTCContext";
+import { useSafePromise } from "flat-components";
 
 export const SpeakerCheckPage = (): React.ReactElement => {
     const { t } = useTranslation();
-    const rtcEngine = useRTCEngine();
+    const rtc = useContext(FlatRTCContext);
     const [devices, setDevices] = useState<Device[]>([]);
     const [currentDeviceID, setCurrentDeviceID] = useState<string | null>(null);
-    const [currentVolume, setCurrentVolume] = useState(30);
+    const [currentVolume, setCurrentVolume] = useState(() => rtc.getSpeakerVolume());
     const [isPlaying, setIsPlaying] = useState(false);
     const history = useHistory<DeviceCheckResults>();
     const location = useLocation<DeviceCheckResults | undefined>();
+    const sp = useSafePromise();
 
-    useEffect(() => {
-        setDevices(rtcEngine.getAudioPlaybackDevices() as Device[]);
-
-        const onAudioDeviceStateChanged = (): void => {
-            setDevices(rtcEngine.getAudioPlaybackDevices() as Device[]);
-        };
-
-        rtcEngine.on("audioDeviceStateChanged", onAudioDeviceStateChanged);
-
-        return () => {
-            rtcEngine.off("audioDeviceStateChanged", onAudioDeviceStateChanged);
-        };
-        // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [rtcEngine]);
+    const onDeviceIDChanged = useCallback(
+        (deviceID: string): void => {
+            console.log("set device", deviceID);
+            rtc.setSpeakerID(deviceID);
+        },
+        [rtc],
+    );
 
     useEffect(() => {
-        if (devices.length !== 0) {
-            setCurrentDeviceID(devices[0].deviceid);
-        }
-    }, [devices]);
+        const updateSpeakerDevices = async (deviceID?: string): Promise<void> => {
+            const devices = await sp(rtc.getSpeakerDevices());
+            setDevices(devices);
+            setCurrentDeviceID(deviceID || devices[0]?.deviceId || null);
+        };
+        updateSpeakerDevices(rtc.getSpeakerID());
+        return rtc.events.on("speaker-changed", updateSpeakerDevices);
+    }, [rtc, sp]);
 
     useEffect(() => {
         if (currentDeviceID && isPlaying) {
-            rtcEngine.setAudioPlaybackDevice(currentDeviceID);
-            rtcEngine.startAudioPlaybackDeviceTest(
+            rtc.startSpeakerTest(
                 path.join(
                     runtime.assetsPath,
+                    "media",
```

Comment:
Have you tested it here?

Reply:
https://github.com/netless-io/flat/tree/main/desktop/renderer-app/src/assets/media